thirdeye-server/src/main/java/ai/startree/thirdeye/auth/AuthorizationManager.java

```diff
@@ -0,0 +1,113 @@
+import com.google.inject.Inject;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.ws.rs.ForbiddenException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+
+public class AuthorizationManager {
+
+  public final AlertManager alertManager;
+  public final AlertTemplateManager alertTemplateManager;
+  public final AccessControl accessControl;
+
+  @Inject
+  public AuthorizationManager(
+      final AlertManager alertManager,
+      final AlertTemplateManager alertTemplateManager,
+      final AccessControl accessControl
+  ) {
+    this.alertManager = alertManager;
+    this.alertTemplateManager = alertTemplateManager;
+    this.accessControl = accessControl;
+  }
+
+  public <T extends AbstractDTO> void ensureCanCreate(final ThirdEyePrincipal principal,
+      final T entity) {
+    ensureHasAccess(principal, ResourceIdentifier.from(entity), AccessType.UPDATE);
+    relatedEntities(entity).forEach(relatedId ->
+        ensureHasAccess(principal, relatedId, AccessType.READ));
+  }
+
+  public <T extends AbstractDTO> void ensureCanDelete(final ThirdEyePrincipal principal,
+      final T entity) {
+    ensureHasAccess(principal, ResourceIdentifier.from(entity), AccessType.UPDATE);
+  }
+
+  public <T extends AbstractDTO> void ensureCanEdit(final ThirdEyePrincipal principal,
+      final T before, final T after) {
+    ensureHasAccess(principal, ResourceIdentifier.from(before), AccessType.UPDATE);
+    ensureHasAccess(principal, ResourceIdentifier.from(after), AccessType.UPDATE);
+    relatedEntities(after).forEach(related ->
+        ensureHasAccess(principal, related, AccessType.READ));
+  }
+
+  public <T extends AbstractDTO> void ensureCanRead(final ThirdEyePrincipal principal,
+      final T entity) {
+    ensureHasAccess(principal, ResourceIdentifier.from(entity), AccessType.READ);
+  }
+
+  public void ensureCanEvaluate(final ThirdEyePrincipal principal, final AlertDTO entity) {
+    ensureCanCreate(principal, entity);
+  }
+
+  public void ensureCanValidate(final ThirdEyePrincipal principal, final AlertDTO entity) {
+    ensureCanCreate(principal, entity);
+  }
+
+  public void ensureHasAccess(final ThirdEyePrincipal principal,
+      final ResourceIdentifier identifier, final AccessType accessType) {
+    if (!hasAccess(principal, identifier, accessType)) {
+      throw forbiddenExceptionFor(principal, identifier, accessType);
+    }
+  }
+
+  public boolean hasAccess(final ThirdEyePrincipal principal,
+      final ResourceIdentifier identifier, final AccessType accessType) {
+    return accessControl.hasAccess(principal, identifier, accessType);
+  }
+
+  private <T extends AbstractDTO> List<ResourceIdentifier> relatedEntities(T entity) {
+    if (entity instanceof AlertDTO) {
+      final AlertDTO alertDto = (AlertDTO) entity;
+      final AlertTemplateDTO alertTemplateDTO = new AlertTemplateRenderer(
```

Comment:
what is the reason for creating a templateRenderer instead of using the singleton?